// pg-adapter/build/adapter.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.Adapter = exports.PgError = exports.ResultMode = exports.AdapterBase = exports.Transaction = exports.parseUrl = exports.raw = exports.quote = void 0;
var types_1 = require("./types");
exports.ResultMode = types_1.ResultMode;
var error_1 = require("./lib/error");
exports.PgError = error_1.PgError;
var parseUrl_1 = require("./lib/parseUrl");
var connect_1 = require("./lib/connect");
var sync_1 = require("./lib/sync");
var close_1 = require("./lib/close");
var defaultDecodeTypes_1 = require("./lib/defaultDecodeTypes");
var adapterBase_1 = require("./lib/adapterBase");
exports.AdapterBase = adapterBase_1.AdapterBase;
var transaction_1 = require("./lib/transaction");
exports.Transaction = transaction_1.Transaction;
var log_1 = require("./lib/log");
var prepare_1 = require("./lib/prepare");
var quote_1 = require("./lib/quote");
__createBinding(exports, quote_1, "quote");
__createBinding(exports, quote_1, "raw");
var parseUrl_2 = require("./lib/parseUrl");
__createBinding(exports, parseUrl_2, "parseUrl");
var Adapter = /** @class */ (function (_super) {
    __extends(Adapter, _super);
    function Adapter(_a) {
        var _b = _a === void 0 ? {} : _a, _c = _b.host, host = _c === void 0 ? '127.0.0.1' : _c, _d = _b.port, port = _d === void 0 ? 5432 : _d, _e = _b.database, database = _e === void 0 ? 'postgres' : _e, _f = _b.user, user = _f === void 0 ? process.env.USER || 'postgres' : _f, _g = _b.password, password = _g === void 0 ? '' : _g, _h = _b.pool, pool = _h === void 0 ? 10 : _h, _j = _b.log, log = _j === void 0 ? Adapter.defaultLog : _j, decodeTypes = _b.decodeTypes;
        var _this = _super.call(this, { pool: pool, decodeTypes: decodeTypes || defaultDecodeTypes_1.defaultDecodeTypes, log: log }) || this;
        _this.connected = false;
        _this.sync = function () { return sync_1.sync(_this); };
        _this.close = function () { return close_1.close(_this); };
        _this.connectionSettings = {
            host: host,
            port: port,
            database: database,
            user: user,
            password: password
        };
        _this.pool = pool;
        return _this;
    }
    Adapter.fromURL = function (urlOrOptions, options) {
        if (typeof urlOrOptions === 'object')
            return new this(__assign(__assign({}, parseUrl_1.parseUrl(process.env.DATABASE_URL)), urlOrOptions));
        else
            return new this(__assign(__assign({}, parseUrl_1.parseUrl(urlOrOptions)), options));
    };
    Adapter.prototype.connect = function () {
        return __awaiter(this, void 0, void 0, function () {
            var promises, i, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (this.connected)
                            return [2 /*return*/];
                        this.connected = true;
                        promises = [];
                        for (i = 0; i < this.pool; i++)
                            promises.push(connect_1.connect(this, this.sockets[i], this.connectionSettings));
                        _a = this;
                        return [4 /*yield*/, Promise.all(promises)];
                    case 1:
                        _a.sockets = _b.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    Adapter.prototype.transaction = function (fn) {
        var error = new error_1.PgError();
        return transaction_1.transaction(this, error, fn);
    };
    Adapter.prototype.wrapperTransaction = function (target, fn) {
        var error = new error_1.PgError();
        return transaction_1.wrapperTransaction(this, error, target, fn);
    };
    Adapter.prototype.prepare = function (_a) {
        var name = _a.name, args = _a.args, query = _a.query;
        return prepare_1.prepare(this, name, args, query);
    };
    Adapter.defaultLog = log_1.defaultLog;
    return Adapter;
}(adapterBase_1.AdapterBase));
exports.Adapter = Adapter;
