// pg-adapter/build/lib/buffer.js
"use strict";
exports.__esModule = true;
exports.getMessage = exports.throwError = exports.noop = exports.skipMessage = exports.getMessageLength = exports.isPositiveInt16 = exports.decodeInt16 = exports.decodeInt32 = exports.encodeInt32 = void 0;
var encodeInt32 = function (buf, i, n) {
    buf[i] = (n >> 24) & 0xff;
    buf[i + 1] = (n >> 16) & 0xff;
    buf[i + 2] = (n >> 8) & 0xff;
    buf[i + 3] = n & 0xff;
};
exports.encodeInt32 = encodeInt32;
var decodeInt32 = function (data, i) {
    return data[i + 3] + (data[i + 2] << 8) + (data[i + 1] << 16) + (data[i] << 24);
};
exports.decodeInt32 = decodeInt32;
var decodeInt16 = function (data, i) {
    return data[i + 1] + (data[i] << 8);
};
exports.decodeInt16 = decodeInt16;
var isPositiveInt16 = function (data, i) {
    return data[i] !== 255;
};
exports.isPositiveInt16 = isPositiveInt16;
var getMessageLength = function (data, pos) {
    return exports.decodeInt32(data, pos + 1);
};
exports.getMessageLength = getMessageLength;
var skipMessage = function (data, pos) {
    return pos + exports.getMessageLength(data, pos) + 1;
};
exports.skipMessage = skipMessage;
var noop = function () {
    // noop
};
exports.noop = noop;
var throwError = function (err) {
    throw err;
};
exports.throwError = throwError;
var getMessage = function (data, pos, offset) {
    if (pos === void 0) { pos = 0; }
    if (offset === void 0) { offset = 0; }
    return data.slice(pos + offset + 5, exports.getMessageLength(data, pos) + 1);
};
exports.getMessage = getMessage;
