// pg-adapter/build/lib/messageHandler.js
"use strict";
exports.__esModule = true;
exports.removeListener = exports.handleMessage = void 0;
var buffer_1 = require("./buffer");
var auth_1 = require("./handlers/auth");
var complete_1 = require("./handlers/complete");
var errorOrNotice_1 = require("./handlers/errorOrNotice");
var parseDescription_1 = require("./handlers/parseDescription");
var parseRow_1 = require("./handlers/parseRow");
var codes;
(function (codes) {
    codes[codes["authenticationCode"] = 'R'.charCodeAt(0)] = "authenticationCode";
    codes[codes["backendKeyDataCode"] = 'K'.charCodeAt(0)] = "backendKeyDataCode";
    codes[codes["parameterStatusCode"] = 'S'.charCodeAt(0)] = "parameterStatusCode";
    codes[codes["readyForQueryCode"] = 'Z'.charCodeAt(0)] = "readyForQueryCode";
    codes[codes["rowDescriptionCode"] = 'T'.charCodeAt(0)] = "rowDescriptionCode";
    codes[codes["dataRowCode"] = 'D'.charCodeAt(0)] = "dataRowCode";
    codes[codes["commandCompleteCode"] = 'C'.charCodeAt(0)] = "commandCompleteCode";
    codes[codes["errorResponseCode"] = 'E'.charCodeAt(0)] = "errorResponseCode";
    codes[codes["noticeResponseCode"] = 'N'.charCodeAt(0)] = "noticeResponseCode";
    codes[codes["bindCode"] = 'B'.charCodeAt(0)] = "bindCode";
    codes[codes["parseCompleteCode"] = '1'.charCodeAt(0)] = "parseCompleteCode";
    codes[codes["bindCompleteCode"] = '2'.charCodeAt(0)] = "bindCompleteCode";
    codes[codes["closeCompleteCode"] = '3'.charCodeAt(0)] = "closeCompleteCode";
    codes[codes["copyDataCode"] = 'd'.charCodeAt(0)] = "copyDataCode";
    codes[codes["copyDoneCode"] = 'c'.charCodeAt(0)] = "copyDoneCode";
    codes[codes["copyFailCode"] = 'f'.charCodeAt(0)] = "copyFailCode";
    codes[codes["copyInResponseCode"] = 'G'.charCodeAt(0)] = "copyInResponseCode";
    codes[codes["copyOutResponseCode"] = 'H'.charCodeAt(0)] = "copyOutResponseCode";
    codes[codes["copyBothResponseCode"] = 'W'.charCodeAt(0)] = "copyBothResponseCode";
    codes[codes["functionCallResponseCode"] = 'V'.charCodeAt(0)] = "functionCallResponseCode";
    codes[codes["negotiateProtocolVersionCode"] = 'v'.charCodeAt(0)] = "negotiateProtocolVersionCode";
    codes[codes["noDataCode"] = 'n'.charCodeAt(0)] = "noDataCode";
    codes[codes["notificationResponseCode"] = 'A'.charCodeAt(0)] = "notificationResponseCode";
    codes[codes["emptyQueryResponseCode"] = 'I'.charCodeAt(0)] = "emptyQueryResponseCode";
    codes[codes["parameterDescriptionCode"] = 't'.charCodeAt(0)] = "parameterDescriptionCode";
    codes[codes["portalSuspendedCode"] = 's'.charCodeAt(0)] = "portalSuspendedCode";
})(codes || (codes = {}));
// TODO: requires proper testing with edge cases
var listener = function (socket, message, creds, data, size) {
    if (size === void 0) { size = data.length; }
    var pos = 0;
    var len = message.cutMessageLength;
    if (len !== 0) {
        // calculate length if it didn't fit in prev block
        if (len === -1) {
            data.copy(message.buffer, message.cutMessageAllocated, 0, 5 - message.cutMessageAllocated);
            len = buffer_1.getMessageLength(message.buffer, 0);
            message.cutMessageLength = len;
            if (message.buffer.length < len)
                message.buffer = Buffer.alloc(len + 1, message.buffer);
        }
        if (len > message.cutMessageAllocated + data.length) {
            data.copy(message.buffer, message.cutMessageAllocated);
            message.cutMessageAllocated += data.length;
            return;
        }
        var copySize = len - message.cutMessageAllocated;
        data.copy(message.buffer, message.cutMessageAllocated);
        message.cutMessageLength = 0;
        message.cutMessageAllocated = 0;
        listener(socket, message, creds, message.buffer, len);
        pos = copySize + 1;
    }
    var task = socket.task;
    while (pos < size) {
        len = pos + 5 < size ? buffer_1.getMessageLength(data, pos) : -1;
        if (len === -1 || pos + len + 1 > size) {
            if (message.buffer.length < len)
                message.buffer = Buffer.alloc(len + 1, message.buffer);
            data.copy(message.buffer, message.cutMessageAllocated, pos);
            message.cutMessageAllocated = size - pos;
            message.cutMessageLength = len;
            break;
        }
        var code = data[pos];
        if (code === codes.dataRowCode) {
            parseRow_1.parseRow(socket, task, data, pos);
        }
        else if (code === codes.rowDescriptionCode) {
            parseDescription_1.parseDescription(socket, task, data, pos);
        }
        else if (code === codes.readyForQueryCode) {
            return task.finish(socket, task);
        }
        else if (code === codes.errorResponseCode) {
            var error = errorOrNotice_1.parseError(task, data, pos);
            task.failed = true;
            Object.assign(task.error, error);
            if (error.level !== 'ERROR')
                return task.finish(socket, task);
        }
        else if (code === codes.noticeResponseCode) {
            var notice = errorOrNotice_1.parseError(task, data, pos);
            if (!task.notices)
                task.notices = [notice];
            else
                task.notices.push(notice);
        }
        else if (code === codes.authenticationCode) {
            auth_1.auth(socket, task, creds, data, pos);
        }
        else if (code === codes.commandCompleteCode) {
            complete_1.complete(task);
        }
        else {
            if (code !== codes.parameterStatusCode &&
                code !== codes.backendKeyDataCode) {
                console.warn("Handling of " + code + " code is not implemented yet");
            }
        }
        pos += len + 1;
    }
};
var handleMessage = function (socket, creds) {
    var message = {
        buffer: Buffer.alloc(10000),
        cutMessageLength: 0,
        cutMessageAllocated: 0
    };
    socket.dataListener = function (data) {
        listener(socket, message, creds, data);
    };
    socket.on('data', socket.dataListener);
};
exports.handleMessage = handleMessage;
var removeListener = function (socket) {
    if (socket.dataListener) {
        socket.removeListener('data', socket.dataListener);
        delete socket.dataListener;
    }
};
exports.removeListener = removeListener;
