// pg-adapter/build/lib/parseUrl.js
"use strict";
exports.__esModule = true;
exports.parseUrl = void 0;
var url_1 = require("url");
var parseUrl = function (url) {
    var _a, _b, _c, _d;
    if (url === void 0) { url = process.env.DATABASE_URL; }
    if (!url)
        throw new Error('Provide url or DATABASE_URL env variable');
    var dbUrl = url_1.parse(url);
    var scheme = ((_a = dbUrl.protocol) === null || _a === void 0 ? void 0 : _a.substr(0, dbUrl.protocol.length - 1)) || '';
    if (scheme.indexOf('postgres') === -1)
        throw new Error("Database url " + url + " does not seem to be postgres");
    var _e = ((_b = dbUrl.host) === null || _b === void 0 ? void 0 : _b.split(':')) || ['', ''], host = _e[0], port = _e[1];
    var _f = ((_c = dbUrl.auth) === null || _c === void 0 ? void 0 : _c.split(':')) || ['', ''], user = _f[0], password = _f[1];
    return {
        host: host,
        port: +port || 5432,
        database: ((_d = dbUrl.path) === null || _d === void 0 ? void 0 : _d.slice(1)) || '',
        user: user,
        password: password
    };
};
exports.parseUrl = parseUrl;
